lib/openhab/core/items/item.rb

```diff
@@ -28,6 +28,35 @@ def ===(other)
         #   The item's descriptive label.
         #   @return [String, nil]
 
+        #
+        # @!attribute [rw] tags
+        #   The item's tags
+        #   @return [Array<String>]
+        #   @overload tags
+        #     Returns the item's tags.
+        #     @return [Array<String>]
+        #   @overload tags=(tags)
+        #     Sets the item's tags.
+        #
+        #     To remove all tags, assign an empty array.
+        #     @param [Array<String,Symbol,Semantics::Tag>] tags Tags to set.
+        #     @return [void]
+        #     @raise [FrozenError] if the provider is not a
+        #       {org.openhab.core.common.registry.ManagedProvider ManagedProvider} that can be updated.
+
+        # @!visibility private
+        def tags=(tags)
+          provider = Provider.registry.provider_for(name)
+          unless provider.is_a?(org.openhab.core.common.registry.ManagedProvider)
+            raise FrozenError, "Cannot modify item #{name} from provider #{provider.inspect}."
+          end
+
+          tags = DSL::Items::ItemBuilder.normalize_tags(*tags)
+          remove_all_tags
```

Comment:
while we're here, can you make (remove_all_tags, removeAllTags, add_tag, addTags, add_tags, addTags) private, so that the provider check and update can't be circumvented?

if you want to be ambitious, you could re-implement (remove_all_tags, add_tag, and add_tags) to do the provider check and update (calling removeAllTags, addTag, and addTags to get the super version).

Reply:
no, not at all. core can't do them because they need to be available for providers to implement their own items. if core adds anything, it would be some other service to update tags, that then does the provider check and update. but we're presenting an interface here for rules, not providers.

Reply:
Isn't the item builder using this interface too? Say if we were to make add_tag private, won't it affect this 
https://github.com/ccutrer/openhab-jrubyscripting/blob/b7d0cc5c4b169f94ebe324aeee81a531bb81f299/lib/openhab/dsl/items/builder.rb#L434

Reply:
Artificially introducing such restrictions, means someone can't later on come and create/implement another item provider using jruby because the interface was only meant for rules.

IMO, let rules be adults, and if they do something they're not supposed to do, but allowed by core, so be it. It's not our job to police everything.